### PR TITLE
Feature/test attach

### PR DESCRIPTION
**Pull request by xross**

### CODE REVIEW

File: tests/usb_phy.py [hunk 1]
```diff
@@ -155,7 +191,6 @@ def run(self):
         self.start_test()
 
         for i, event in enumerate(self._session.events):
-
             event.drive(self, self._session.bus_speed)
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_phy.py:194]: Do we need `enumerate` here? Looks like `i` has been factored out, could just be:
```
    for event in self._sesion.bus_speed:
        event.drive(self, self._session.bus_speed)
```
Also, is `drive` ever called with a bus speed that is anything other than `self._session.bus_speed`? Could just de-reference 'self' inside the function, save the cost of argument marshalling? (Although obv. if there's any chance this needs to change, keep it as a parameter, always....)

File: tests/usb_signalling.py [hunk 2]
```diff
@@ -5,6 +5,162 @@
...
+                print("ERROR: DUT didnt not assert XcvrSel & TermSel quickly enough")
+
+            # Check device asserts XcvrSel and TermSel before T_SIGATT
+            xcvrsel = xsi.sample_periph_pin(usb_phy._xcvrsel)
+            termsel = xsi.sample_periph_pin(usb_phy._termsel)
+
+            if xcvrsel == 1 and termsel == 1:
+                break
+
+        print("DUT entered FS")
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_signalling.py:62]: Is this correct? We pass in bus speed as a parameter at the top, why is it unconditionally 'FS' here? (Unless this isn't bus speed, in which case apologies ;-))

File: tests/usb_signalling.py [hunk 3]
```diff
@@ -5,6 +5,162 @@
...
+
+            # Ensure DUT enters HS before T0 + T_DRST
+
+            # Drive HS Idle (SE0) on bus
+            xsi.drive_periph_pin(usb_phy._ls, USB_LINESTATE["IDLE"])
+
+            # TODO how long to drive SE0 for?
+            wait_until_ns(time() + 10000)
+
+        return
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_signalling.py:161]: Why return from a void function?

**Review by DaveAtkinsonXMOS — COMMENTED** [review]
Nothing I couldn't give a tick for if pushed, let me know.

File: lib_xud/module_build_info [hunk 4]
```diff
@@ -16,7 +16,7 @@ XCC_FLAGS_dfu_flash.xc = $(MODULE_XCC_FLAGS) -Os
 XCC_FLAGS_XUD_Client.xc = $(MODULE_XCC_FLAGS) -mno-dual-issue
 XCC_FLAGS_XUD_Main.xc = $(MODULE_XCC_FLAGS) -mno-dual-issue
 XCC_FLAGS_XUD_SetDevAddr.xc = $(MODULE_XCC_FLAGS) -mno-dual-issue
-XCC_FLAGS_XUD_DeviceAttach.xc = $(MODULE_XCC_FLAGS) -mno-dual-issue -Wno-return-type
+#XCC_FLAGS_XUD_DeviceAttach.xc = $(MODULE_XCC_FLAGS) -mno-dual-issue -Wno-return-type
```
[the remark this answers is not in the document]
Comment (xross) [reply lib_xud/module_build_info:19]: Done

File: tests/test_device_attach/src/main.xc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    par
+    {
+        { 
+            #if defined(XUD_TEST_SPEED_FS)
+            unsigned speed = XUD_SPEED_FS;
+            #elif defined(XUD_TEST_SPEED_HS)
+            unsigned speed = XUD_SPEED_HS;
+            #else
+            #error XUD_TEST_SPEED_XX not defined
+            #endif
```
[the remark this answers is not in the document]
Comment (xross) [reply tests/test_device_attach/src/main.xc:47]: Done

File: tests/usb_phy.py (hunk 1 above)
Comment (xross) [reply tests/usb_phy.py:194]: I guess at some point I was used for something... removed.

Comment (xross) [reply tests/usb_phy.py:194]: In future we might want to send other values, but I'll fix it for now.

Comment (xross) [reply tests/usb_phy.py:194]: Actually, eventually Id like to remove passing usb_phy to drive so I think I'll leave it as is for now

File: tests/usb_signalling.py (hunk 2 above)
Comment (xross) [reply tests/usb_signalling.py:62]: All signalling is done at FS, so a HS device must switch back to FS when suspended etc.

File: tests/usb_signalling.py (hunk 3 above)
Comment (xross) [reply tests/usb_signalling.py:161]: removed

**Review by DaveAtkinsonXMOS — APPROVED** [review]
It's a yes from me...

**Pull request merged**